jbpm-benchmarks/kieserver-performance-tests/src/main/java/org/jbpm/test/performance/taskassigning/updates/TaskAssigningUpdates.java

```diff
@@ -0,0 +1,228 @@
+        startProcesses(processCount);
+
+        Set<Long> tasksInProgress = new ConcurrentHashMap<>().newKeySet();
+        // Keep completing tasks to introduce changes and trigger incremental re-planning.
+        while (completedTasks.size() < taskCount) {
+            TaskQueryFilterSpec filterSpec = new TaskQueryFilterSpec();
+            List<TaskInstance> assignedTasks = getQueryClient()
+                    .findHumanTasksWithFilters(ASSIGNED_TASKS_QUERY_NAME, filterSpec, 0, userCount);
+            if (assignedTasks.isEmpty()) { // If there are no tasks to work on, wait and retry.
+                sleep(500L);
+                continue;
+            }
+            LOGGER.debug(String.format("Completing %d tasks. Already finished %d tasks of %d.",
+                    assignedTasks.size(), completedTasks.size(), taskCount));
+            // Run each task in a separate thread.
+            for (TaskInstance task : assignedTasks) {
+                // Avoid submitting task some other thread already takes care of
+                if (!tasksInProgress.contains(task.getId())) {
+                    tasksInProgress.add(task.getId());
+                    CompletableFuture.runAsync(() -> {
+                        getTaskClient().startTask(CONTAINER_ID, task.getId(), task.getActualOwner());
+                        sleep(TASK_COMPLETION_DELAY_MILLIS);
+                        getTaskClient().completeTask(CONTAINER_ID, task.getId(), task.getActualOwner(), new HashMap<>());
+                        completedTasks.add(task.getId());
+                        tasksInProgress.remove(task.getId());
+                    }, threadPoolExecutor);
+                }
+            }
+        }
+
+        Map<String, NavigableSet<TaskEventInstance>> taskEventsPerUser = getTaskEventsPerUser();
+        List<Long> allDelaysBetweenCompleteAndStartEvents = taskEventsPerUser.entrySet().stream()
+                .map((entry) -> delaysBetweenCompleteAndStartEvents(entry.getValue()))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        long median = median(allDelaysBetweenCompleteAndStartEvents);
+        taskAssignmentDuration.update(median, TimeUnit.MILLISECONDS);
+
+        afterScenario();
+    }
+
+    private void afterScenario() {
+        shutdownExecutorService(threadPoolExecutor);
+        abortAllProcesses();
+    }
+
+    /**
+     * Computes delays between completing and starting consecutive tasks.
+     */
+    private List<Long> delaysBetweenCompleteAndStartEvents(SortedSet<TaskEventInstance> events) {
+        List<Long> delays = new ArrayList<>();
+        TaskEventInstance previous = null;
+        for (TaskEventInstance taskEventInstance : events) {
+            if (TASK_COMPLETED.equals(taskEventInstance.getType())) {
+                previous = taskEventInstance;
+            } else if (previous != null && TASK_STARTED.equals(previous.getType())) {
```

Comment:
I guess here it should be `TASK_STARTED.equals(taskEventInstance.getType())`? As you never assign `TASK_STARTED` type to the previous.

Reply:
Nice catch! Thank you.